go/chat/attachment_s3.go

```diff
@@ -0,0 +1,341 @@
+		if err != nil && err != io.ErrUnexpectedEOF && err != io.EOF {
+			return err
+		}
+		if n < blockSize {
+			block = block[:n]
+		}
+		if n > 0 {
+			md5sum := md5.Sum(block)
+			md5hex := hex.EncodeToString(md5sum[:])
+
+			if previous != nil {
+				// resuming an upload, so check local stash record
+				// and abort on mismatch before adding a job for this block
+				lhash, found := previous.Parts[partNumber]
+				if found && lhash != md5hex {
+					a.log.Debug("part %d failed local part record verification", partNumber)
+					return ErrAbortOnPartMismatch
+				}
+			}
+
+			if err := a.addJob(ctx, blockCh, block, partNumber, md5hex); err != nil {
+				return err
+			}
+		}
+		if err == io.EOF || err == io.ErrUnexpectedEOF {
+			break
+		}
+
+		if a.blockLimit > 0 && partNumber >= a.blockLimit {
+			a.log.Debug("hit blockLimit of %d", a.blockLimit)
+			break
+		}
+	}
+	return nil
+}
+
+// addJob creates a job and puts it on blockCh, unless the blockCh isn't ready and the context has been canceled.
+func (a *AttachmentStore) addJob(ctx context.Context, blockCh chan job, block []byte, partNumber int, hash string) error {
+	// Create a job, unless the context has been canceled.
+	select {
+	case blockCh <- job{block: block, index: partNumber, hash: hash}:
+		return nil
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+}
+
+// uploadPart handles uploading a job to S3.  The job `b` has already passed local stash verification.
+// If this is a resumed upload, it checks the previous parts reported by S3 and will skip uploading
+// any that already exist.
+func (a *AttachmentStore) uploadPart(ctx context.Context, task *UploadTask, b job, previous *AttachmentInfo, previousParts map[int]s3.Part, multi s3.MultiInt, retCh chan s3.Part) error {
+	a.log.Debug("start: upload part %d", b.index)
+
+	// check to see if this part has already been uploaded.
+	// for job `b` to be here, it has already passed local stash verification.
+	if previous != nil {
+		// check s3 previousParts for this block
+		p, ok := previousParts[b.index]
+		if ok && int(p.Size) == len(b.block) && p.ETag == b.etag() {
+			a.log.Debug("part %d already uploaded to s3", b.index)
+
+			// part already uploaded, so put it in the retCh unless the context
+			// has been canceled
+			select {
+			case retCh <- p:
+			case <-ctx.Done():
+				return ctx.Err()
+			}
+
+			// nothing else to do
+			return nil
+		}
+
+		if p.Size > 0 {
+			// only abort if the part size from s3 is > 0.
+			a.log.Debug("part %d s3 mismatch:  size %d != expected %d or etag %s != expected %s", b.index, p.Size, len(b.block), p.ETag, b.etag())
+			return ErrAbortOnPartMismatch
+		}
+
+		// this part doesn't exist on s3, so it needs to be uploaded
+		a.log.Debug("part %d not uploaded to s3 by previous upload attempt", b.index)
+	}
+
+	// stash part info locally before attempting S3 put
+	if err := a.stash.RecordPart(task.stashKey(), b.index, b.hash); err != nil {
+		a.log.Debug("StashRecordPart error: %s", err)
+	}
```

Comment:
Checking the stash only happens during resume attempts, RecordPart happens with every upload for resuming.

We haven't tried a put yet during makeBlockJobs, it's just in the pipeline.  

If you feel strongly, I can move it.  I like the locality of the call here, but I understand your preference as well.